File: client/pagination.go
package client

import (
	"context"
	"iter"
	"net/url"
	"strconv"
)

// PageInfo describes a page's position in the full result set. TotalCount is
// present only when the request asked for it (ListOptions.Total).
type PageInfo struct {
	// HasNextPage reports that NextCursor will fetch more.
	HasNextPage bool `json:"has_next_page"`

	// HasPreviousPage reports only that this is not the first page. It is
	// not a backward-paging capability: the API has no "before" cursor and
	// no direction, so there is nothing to call to go back.
	//
	// To offer a Back button, keep the cursors you have already used and
	// re-request one — CursorStack does exactly that.
	HasPreviousPage bool `json:"has_previous_page"`

	NextCursor *string `json:"next_cursor,omitempty"`
	TotalCount *int    `json:"total_count,omitempty"`
}

// CursorStack remembers the cursor of each page a caller has visited, so a
// paged UI can step backwards over a forward-only keyset API.
//
// PageInfo.HasPreviousPage says a previous page exists but the API offers no
// way to ask for it, so every paged UI has to keep this state itself. The
// admin console does, in its own composable; this is the same thing for Go
// callers.
//
// A CursorStack is not safe for concurrent use: it belongs to one view.
//
//	var st client.CursorStack
//	page, err := c.Types().List(ctx, client.ListOptions{Cursor: st.Current()})
//	// forward:
//	st.Push(page.PageInfo)
//	// backward:
//	st.Pop()
type CursorStack struct {
	// visited holds the cursor used for each page in order. The first entry
	// is "" — the first page takes no cursor.
	visited []string
}

// Current returns the cursor for the page the stack is on. It is "" on the
// first page, which is what a list call wants there.
func (s *CursorStack) Current() string {
	if len(s.visited) == 0 {
		return ""
	}
	return s.visited[len(s.visited)-1]
}

// Push records that the caller moved forward, and returns the cursor for the
// next page. It reports false when info has no next page, leaving the stack
// untouched — so a Next button at the end of a list is a no-op rather than an
// empty page.
func (s *CursorStack) Push(info PageInfo) (string, bool) {
	if !info.HasNextPage || info.NextCursor == nil || *info.NextCursor == "" {
		return s.Current(), false
	}
	if len(s.visited) == 0 {
		s.visited = append(s.visited, "")
	}
	s.visited = append(s.visited, *info.NextCursor)
	return s.Current(), true
}

// Pop steps back one page and returns that page's cursor. It reports false on
// the first page, where there is nothing to go back to.
func (s *CursorStack) Pop() (string, bool) {
	if len(s.visited) < 2 {
		return "", false
	}
	s.visited = s.visited[:len(s.visited)-1]
	return s.Current(), true
}

// Depth is the zero-based index of the page the stack is on.
func (s *CursorStack) Depth() int {
	if len(s.visited) == 0 {
		return 0
	}
	return len(s.visited) - 1
}

// Page is one page of a list response.
type Page[T any] struct {
	Items    []T      `json:"items"`
	PageInfo PageInfo `json:"page_info"`
}

// ListOptions are the pagination arguments common to every list call. The
// cursor is opaque and keyset-based, so pages stay stable under concurrent
// writes; the total is computed only when Total is set.
type ListOptions struct {
	Limit  int    // page size; 0 uses the server default
	Cursor string // opaque keyset cursor; "" starts at the first page
	Total  bool   // request PageInfo.TotalCount
}

func (o ListOptions) apply(q url.Values) {
	if o.Limit > 0 {
		q.Set("limit", strconv.Itoa(o.Limit))
	}
	if o.Cursor != "" {
		q.Set("cursor", o.Cursor)
	}
	if o.Total {
		q.Set("total", "true")
	}
}

func firstOpts(opts []ListOptions) ListOptions {
	if len(opts) > 0 {
		return opts[0]
	}
	return ListOptions{}
}

// paginate yields every item across all pages by following the keyset cursor.
// A fetch error is yielded once (with the zero item) and ends iteration; break
// out of the range loop to stop early.
func paginate[T any](fetch func(cursor string) (*Page[T], error)) iter.Seq2[T, error] {
	return func(yield func(T, error) bool) {
		cursor := ""
		for {
			page, err := fetch(cursor)
			if err != nil {
				var zero T
				yield(zero, err)
				return
			}
			for _, it := range page.Items {
				if !yield(it, nil) {
					return
				}
			}
			if !page.PageInfo.HasNextPage || page.PageInfo.NextCursor == nil {
				return
			}
			cursor = *page.PageInfo.NextCursor
		}
	}
}

// listPage is the shared one-page GET for a keyset list endpoint.
func listPage[T any](ctx context.Context, c *Client, path string, base url.Values, opts ListOptions) (*Page[T], error) {
	q := cloneValues(base)
	opts.apply(q)
	var out Page[T]
	if err := c.do(ctx, "GET", path, q, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// items GETs a non-paginated {"items":[...]} list.
func items[T any](ctx context.Context, c *Client, path string, q url.Values) ([]T, error) {
	var out struct {
		Items []T `json:"items"`
	}
	if err := c.do(ctx, "GET", path, q, nil, &out); err != nil {
		return nil, err
	}
	return out.Items, nil
}

func cloneValues(v url.Values) url.Values {
	out := url.Values{}
	for k, vals := range v {
		for _, val := range vals {
			out.Add(k, val)
		}
	}
	return out
}

File: client-ts/src/pagination.ts
import { FlexitypeError } from './errors.js'
import type { RequestOptions } from './http.js'

/**
 * Where a page sits in the whole result set.
 *
 * `has_previous_page` reports only that this is not the first page. It is not
 * a backward-paging capability: the API has no "before" cursor and no
 * direction. To offer a Back button, keep the cursors you have already used —
 * `CursorStack` does exactly that.
 */
export interface PageInfo {
  has_next_page?: boolean
  has_previous_page?: boolean
  next_cursor?: string
  /** Present only when the request asked for it with `total: true`. */
  total_count?: number
}

/** One page of a list response. */
export interface Page<T> {
  items: T[]
  page_info: PageInfo
}

/**
 * The pagination arguments every list call accepts.
 *
 * The cursor is an opaque keyset cursor, so pages stay stable while other
 * writers work. A cursor the ordering column cannot parse, or one carrying the
 * wrong number of values, answers 422 — a VALIDATION FlexitypeError, which the
 * retry policy never replays.
 */
export interface ListOptions extends RequestOptions {
  /** The page size. Omit it for the server default; the maximum is 500. */
  limit?: number | undefined
  /** The opaque cursor of the page to fetch. Omit it for the first page. */
  cursor?: string | undefined
  /** Asks the service to compute `page_info.total_count`. It costs a second scan. */
  total?: boolean | undefined
}

/** Fetches one page given a cursor. It is what the iterators walk. */
export type PageFetcher<T> = (cursor: string | undefined) => Promise<Page<T>>

/**
 * Walks every page by following the keyset cursor.
 *
 * The walk stops when the service reports no next page. It also stops if the
 * service ever returns the cursor it was given, which would otherwise loop for
 * ever on the same page.
 */
export async function* paginatePages<T>(fetchPage: PageFetcher<T>): AsyncGenerator<Page<T>, void, undefined> {
  let cursor: string | undefined
  const seen = new Set<string>()
  for (;;) {
    const page = await fetchPage(cursor)
    yield page
    const next = page.page_info?.next_cursor
    if (page.page_info?.has_next_page !== true || next === undefined || next === '') return
    if (seen.has(next)) {
      throw new FlexitypeError({
        code: 'INTERNAL',
        message: 'flexitype: the service repeated a pagination cursor; the walk was stopped',
        status: 0,
      })
    }
    seen.add(next)
    cursor = next
  }
}

/** Walks every item across every page. */
export async function* paginate<T>(fetchPage: PageFetcher<T>): AsyncGenerator<T, void, undefined> {
  for await (const page of paginatePages(fetchPage)) {
    for (const item of page.items) yield item
  }
}

/**
 * Drains an async iterable into an array.
 *
 * It loads every page, so bound it with an FQL filter or a limit before using
 * it on a type that holds millions of entities.
 */
export async function collect<T>(source: AsyncIterable<T>): Promise<T[]> {
  const out: T[] = []
  for await (const item of source) out.push(item)
  return out
}

/**
 * Remembers the cursor of each page a caller has visited, so a paged screen
 * can step backwards over a forward-only keyset API. It mirrors `CursorStack`
 * in the Go client.
 *
 * A CursorStack belongs to one view and is not safe to share between two.
 *
 * ```ts
 * const stack = new CursorStack()
 * let page = await client.types.list({ cursor: stack.current() })
 * if (stack.push(page.page_info)) page = await client.types.list({ cursor: stack.current() })
 * if (stack.pop()) page = await client.types.list({ cursor: stack.current() })
 * ```
 */
export class CursorStack {
  /** The cursor of each visited page in order. The first entry is always "". */
  private visited: string[] = []

  /** The cursor for the page the stack is on. It is undefined on page one. */
  current(): string | undefined {
    const top = this.visited[this.visited.length - 1]
    return top === undefined || top === '' ? undefined : top
  }

  /**
   * Records a move forward. It reports false when the page info has no next
   * page and leaves the stack untouched, so a Next button at the end of a list
   * does nothing instead of fetching an empty page.
   */
  push(info: PageInfo): boolean {
    const next = info.next_cursor
    if (info.has_next_page !== true || next === undefined || next === '') return false
    if (this.visited.length === 0) this.visited.push('')
    this.visited.push(next)
    return true
  }

  /** Steps back one page. It reports false on the first page. */
  pop(): boolean {
    if (this.visited.length < 2) return false
    this.visited.pop()
    return true
  }

  /** The zero-based index of the page the stack is on. */
  depth(): number {
    return this.visited.length === 0 ? 0 : this.visited.length - 1
  }

  /** Forgets every visited page, returning to page one. */
  reset(): void {
    this.visited = []
  }
}
